src/reactjs/QueryClientProvider.tsx

```diff
@@ -50,6 +58,7 @@ export const QueryClientProvider: React.FC<QueryClientProviderProps> = ({
   client,
   contextSharing = false,
   children,
+  context,
```

Comment:
from an api perspective - does it make sense to allow passing in `context` _and_ `contextSharing` ? I think this would be an invalid state. One solution would be to either combine them to the same prop, or to make the props a union type where you cannot pass in one when you have already provided the other ....

Reply:
I have attempted to type this with https://github.com/tannerlinsley/react-query/pull/2548/commits/358dfd0aa1441301bc1b018ba6a995ec94827e61. Let me know if this does or does not accomplish what you were thinking.

Reply:
> It shouldn't be an invalid state. context takes priority over contextSharing

I'm aware. I'm just thinking from an api perspective, you could provide props that make no sense. Similar to how:

`<Button isPrimary isSecondary>`

makes no sense and it becomes an implementation detail over what takes precedence, something like:

`<Button variant="primary">` or `<Button variant="secondary">`

makes for a better api.

So here, you can do:

```
<QueryClientProvider context={MyContext} contextSharing>
```

but then yeah contextSharing will be `false` even though you provided `true` as a prop. That is the impossible state :)

---

we could do `context = React.Context | 'sharing'` to unify the props. I'm not sure if I like that or not, but we use such polymorphic props in other places too, like:

`notifyOnChangeProps?: Array<keyof InfiniteQueryObserverResult> | 'all'` 🤔

Reply:
@TkDodo I believe I have already captured this as part of https://github.com/tannerlinsley/react-query/commit/358dfd0aa1441301bc1b018ba6a995ec94827e61. TS will not let you pass both `contextSharing` and `context`. You can see this with:

```typescript
type QueryClientProviderPropsBase = {
  client: QueryClient
}
type QueryClientProviderPropsWithContext = ContextOptions &
  QueryClientProviderPropsBase
type QueryClientProviderPropsWithContextSharing = {
  contextSharing?: boolean
} & QueryClientProviderPropsBase

export type QueryClientProviderProps =
  | QueryClientProviderPropsWithContext
  | QueryClientProviderPropsWithContextSharing
```

Reply:
Right, I missed that, sorry. Yes I think that is one way of ensuring that, even though it's two separate props :)

also sorry there are conflicts again 🙈 . can you update with alpha once more please?

Reply:
PR has been rebased to alpha branch and pushed up. Should be no conflicts now.

Also, `context` and `contextSharing` have been consolidated into the `context` prop for `QueryClientProvider`. Docs should also have been updated to match. See this commit:  https://github.com/tannerlinsley/react-query/pull/2548/commits/0794e1f28ed6b10035929858ad1ffc9779c2ecd4.